File: 0530. Minimum Absolute Difference in BST/Solution.py
from _typeshed import SupportsItemAccess


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def getMinimumDifference_MK1(self, root: TreeNode) -> int:
        vals = []

        def inOrder(root: TreeNode) -> None:
            if root.left:
                inOrder(root.left)
            vals.append(root.val)
            if root.right:
                inOrder(root.right)

        inOrder(root)
        return min(abs(a - b) for a, b in zip(vals, vals[1:]))

    def getMinimumDifference_MK2(self, root: TreeNode) -> int:
        stack = []
        p = root
        pre, ans = None, float('inf')
        while p or stack:
            while p:
                stack.append(p)
                p = p.left
            p = stack.pop()
            if pre:
                ans = min(ans, abs(p.val - pre.val))
            pre = p
            p = p.right
        return ans

    def getMinimumDifference_MK3(self, root: TreeNode) -> int:
        def getSuccessor(root: TreeNode) -> TreeNode:
            p = root.left
            while p.right and p.right != root:
                p = p.right
            return p
        
        ans = float('inf')
        pre = None
        while root:
            if pre:
                ans = min(ans, abs(root.val - pre.val))
            if not root.left:
                pre = root
                root = root.right
            else:
                succ = getSuccessor(root)
                if succ.right:
                    succ.right = None
                    pre = root
                    root = root.right
                else:
                    succ.right = root
                    root = root.left
        return ans

File: 0530. Minimum Absolute Difference in BST/Solution.java
import java.util.Stack;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private int ans = Integer.MAX_VALUE;
    private Integer pre = null;

    private void inOrder(TreeNode root) {
        if (root == null)
            return;
        inOrder(root.left);
        if (pre != null)
            this.ans = Math.min(ans, Math.abs(root.val - pre));
        pre = root.val;
        inOrder(root.right);
    }

    private TreeNode getSuccessor(TreeNode root) {
        TreeNode p = root.left;
        while (p.right != null && p.right != root) {
            p = p.right;
        }
        return p;
    }

    public int getMinimumDifference_MK1(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode node = root, pre = null;
        int ans = Integer.MAX_VALUE;
        while (node != null || !stack.empty()) {
            while (node != null) {
                stack.push(node);
                node = node.left;
            }
            node = stack.pop();
            if (pre != null)
                ans = Math.min(ans, node.val - pre.val);
            pre = node;
            node = node.right;
        }
        return ans;
    }

    public int getMinimumDifference_MK2(TreeNode root) {
        inOrder(root);
        return ans;
    }

    public int getMinimumDifference_MK3(TreeNode root) {
        int res = Integer.MAX_VALUE;
        TreeNode pre = null;
        while (root != null) {
            if (pre != null) {
                res = Math.min(res, Math.abs(root.val - pre.val));
            }
            if (root.left == null) {
                pre = root;
                root = root.right;
            } else {
                TreeNode succ = getSuccessor(root);
                if (succ.right == root) {
                    succ.right = null;
                    pre = root;
                    root = root.right;
                } else {
                    succ.right = root;
                    root = root.left;
                }
            }
        }
        return res;
    }
}
